### PR TITLE
Fix #17290 - Cancel gemini transaction if status is not completed after 3 retries

### PR DESCRIPTION
**Pull request by jumde**
Resolves https://github.com/brave/brave-browser/issues/17290

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

#### Cancellation

1. With a clean profile, run brave with the flag `--rewards=gemini-retries=0`
2. Enable verbose logging and rewards
3. Verify with Gemini
4. Navigate to `jumde.github.io` and send a tip 
5. Navigate to `brave://rewards-internals` and verify that the transaction was successfully cancelled.

#### Successful tips and auto-contribute

1. With a clean profile, start brave with the flag `-rewards=reconcile-interval=5`
2. Enable verbose logging and rewards
3. Verify with Gemini
4. Navigate to `jumde.github.io` and send a tip 
5. Tip should be successful
6. Auto-Contribute should be successful after 5 mins

### CODE REVIEW

File: vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.cc [hunk 1]
```diff
@@ -118,5 +126,29 @@ void GeminiTransfer::OnTransactionStatus(const type::Result result,
...
+  gemini_server_->post_cancel_transaction()->Request(wallet->token, id,
+                                                     url_callback);
+}
+
+void GeminiTransfer::OnTransactionCancelled(
+    const type::Result result,
+    const std::string& id,
+    client::TransactionCallback callback) {
+  // Gemini transaction could not be completed
+  BLOG(1, "Transaction id: " << id << " cancelled");
```
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.cc:149]: Maybe include "Gemini" in this log entry.

File: vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.cc [hunk 2]
```diff
@@ -52,6 +52,8 @@ void GeminiTransfer::OnCreateTransaction(const type::Result result,
     return;
   }
 
+  BLOG(0, "Number of retries: " << ledger::gemini_retries);
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.cc:55]: Looks like the BLOG on line 111 should be fixed as well.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoint/gemini/post_cancel_transaction/post_cancel_transaction_gemini.cc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+#include "bat/ledger/internal/endpoint/gemini/post_cancel_transaction/post_cancel_transaction_gemini.h"
+
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/stringprintf.h"
+#include "bat/ledger/internal/endpoint/gemini/gemini_utils.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
```
[the remark this answers is not in the document]
Comment (jumde) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoint/gemini/post_cancel_transaction/post_cancel_transaction_gemini.cc]: resolved

File: vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.cc [hunk 4]
```diff
@@ -118,5 +126,29 @@ void GeminiTransfer::OnTransactionStatus(const type::Result result,
...
+                                       client::TransactionCallback callback) {
+  auto wallet = ledger_->gemini()->GetWallet();
+  if (!wallet) {
+    BLOG(0, "Wallet is null");
+    callback(type::Result::LEDGER_ERROR, "");
+    return;
+  }
+
+  auto url_callback = std::bind(&GeminiTransfer::OnTransactionCancelled, this,
+                                _1, id, callback);
```
[the remark this answers is not in the document]
Comment (jumde) [reply vendor/bat-native-ledger/src/bat/ledger/internal/gemini/gemini_transfer.cc]: resolved!

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoint/gemini/post_cancel_transaction/post_cancel_transaction_gemini.cc [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+std::string PostCancelTransaction::GetUrl(const std::string& tx_ref) {
+  return GetApiServerUrl(
+      base::StringPrintf("/v1/payment/cancel/%s", tx_ref.c_str()));
+}
+
+void PostCancelTransaction::Request(const std::string& token,
+                                    const std::string& tx_ref,
+                                    PostCancelTransactionCallback callback) {
+  auto url_callback =
+      std::bind(&PostCancelTransaction::OnRequest, this, _1, callback);
```
[the remark this answers is not in the document]
Comment (jumde) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoint/gemini/post_cancel_transaction/post_cancel_transaction_gemini.cc:41]: this requires update to ledger client's `LoadURLCallback` which will eventually require an update to all `LoadURL` calls in `bat-native-ledger`. Addressing this as follow-up issues to: https://github.com/brave/brave-browser/issues/18560 - makes sense. Let me know what you think.

**Pull request merged**